opentelemetry-exporter-geneva/geneva-uploader/src/payload_encoder/otlp_encoder.rs

```diff
@@ -39,50 +39,57 @@ impl OtlpEncoder {
-        I: Iterator<Item = &'a opentelemetry_proto::tonic::logs::v1::LogRecord>,
+        I: IntoIterator<Item = &'a opentelemetry_proto::tonic::logs::v1::LogRecord>,
     {
         use std::collections::HashMap;
 
         let mut batches: LogBatches = HashMap::new();
 
         for log_record in logs {
-            // 1. Get schema
-            let field_specs = self.determine_fields(log_record);
-            let schema_id = Self::calculate_schema_id(&field_specs);
-            let (schema_entry, field_info) = self.get_or_create_schema(schema_id, field_specs);
-
-            // 2. Encode row
-            let row_buffer = self.write_row_data(log_record, &field_info);
+            // 1. Get schema with optimized single-pass field collection and schema ID calculation
             let event_name = if log_record.event_name.is_empty() {
                 "Log".to_string()
             } else {
                 log_record.event_name.clone()
             };
+            let (field_specs, schema_id) =
+                self.determine_fields_and_schema_id(log_record, &event_name);
+
+            let schema_entry = self.get_or_create_schema(schema_id, field_specs.clone());
+
+            // 2. Encode row
+            let row_buffer = self.write_row_data(log_record, &field_specs);
             let level = log_record.severity_number as u8;
 
-            // 3. Insert into batches - Key is (schema_id, event_name)
-            batches
-                .entry((schema_id, event_name.clone()))
-                .or_insert_with(|| (schema_entry, Vec::new()))
-                .1
-                .push((event_name, level, row_buffer));
+            // 3. Create batch key
+            let batch_key = event_name.clone();
+
+            // 4. Create or get existing batch entry
+            let entry = batches
+                .entry(batch_key)
+                .or_insert_with(|| (Vec::new(), Vec::new()));
+
+            // 5. Add schema entry if not already present (multiple schemas per event_name batch)
+            if !entry.0.iter().any(|s| s.id == schema_id) {
+                entry.0.push(schema_entry);
+            }
+
+            // 6. Add encoded row to batch
+            entry.1.push((level, schema_id, row_buffer));
         }
 
-        // 4. Encode blobs (one per schema AND event_name combination)
+        // 4. Encode blobs (one per event_name, potentially multiple schemas per blob)
         let mut blobs = Vec::new();
-        for ((schema_id, batch_event_name), (schema_entry, records)) in batches {
+        for (batch_event_name, (schema_entries, records)) in batches {
             let events: Vec<CentralEventEntry> = records
                 .into_iter()
-                .map(|(event_name, level, row_buffer)| CentralEventEntry {
+                .map(|(level, schema_id, row_buffer)| CentralEventEntry {
                     schema_id,
                     level,
-                    event_name,
+                    event_name: batch_event_name.clone(),
```

Comment:
On some more thought, this is kind of defeating the purpose of removing `event_name` from `EncodedRow` since we are cloning the `event_name` here again anyway. On top of that, we also have a vec allocation just to extend `event_name` to the encoded row entries.

I think the best thing to do here would be:
- Avoid this vec allocation for `CentralEventEntry` by having `BatchValue` store `CentraEventEntry` from the get-go instead of storing `EncodedRow` and creating another vec just to extend `event_name`
- Update the type for `event_name` in `CentralEventEntry` to use `Arc<String>` to avoid repeated allocations when cloning the `event_name`.